app/services/tool_filter_service.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+class ToolFilterService
+  def initialize(params)
+    @params = params
+  end
+
+  def call
+    tool_groups = []
+    ToolGroup.all.each { |tool_group| tool_groups << tool_group if match_params(tool_group, @params) }
+    grouped_array = group_resources(tool_groups)
+    order_results(grouped_array)
+  end
+
+  private
+
+  def order_results(grouped_array)
+    result = grouped_array.map do |key, values|
+      counter = values.size
+      sum = values.sum { |o| o[:tool_group_suggestions_weight] * o[:resource_tool_group_suggestions_weight] }
+      average = sum / counter
+      [Resource.find(key), average]
+    end
+
+    sorted_result = result.sort_by { |subarray| -subarray[1] }
+    sorted_result.map { |subarray| subarray[0] }
+  end
+
+  def group_resources(tool_groups)
+    result = tool_groups.flat_map do |tool_group|
+      tool_group.resource_tool_groups.map do |resource|
+        {
+          tool_group_suggestions_weight: tool_group.suggestions_weight,
+          resource_id: resource.resource_id,
+          resource_tool_group_suggestions_weight: resource.suggestions_weight
+        }
+      end
+    end
+
+    result.group_by { |hash| hash[:resource_id] }
+  end
+
+  def match_params(tool_group, params)
+    country = params["filter"]["country"]&.upcase
+    languages = params["filter"]["languages"]
+    openness = params["filter"]["openness"].to_i
+    confidence = params["filter"]["confidence"].to_i
+
+    return true if no_rules_for(tool_group)
+    return true if language_rule(tool_group, languages)
```

Comment:
I have a tool group defined with the following 2 rules:
- language: en
- praxis: openness: 1 confidence: 3

When i submit the following suggestions query:

```
curl --location --globoff 'https://mobile-content-api-stage.cru.org/resources/suggestions?filter[languages][]=en&filter[openness]=1&filter[confidence]=1'
```

I would expect to not get back any results because the confidence doesn't match, but it is still matching the group and returning results.

Reply:
I believe this logic is where it is failing because `language_rule` has the following statement:
```
return false unless tool_group.rule_praxes.none? || tool_group.rule_countries.none?
```
which still processes as true because there were no defined country rules.

Also, since you have each rule type only being evaluated if a rule exists you probably no longer need the `no_rules_for` or `language_rule` special cases.